tools/explorer/tt_adapter/src/tt_adapter/mlir.py

```diff
@@ -0,0 +1,551 @@
+        )
+    )
+    result.append(
+        graph_builder.KeyValue(key="memref_shape", value=str(layout.memref.shape))
+    )
+    result.append(
+        graph_builder.KeyValue(key="memref_rank", value=str(layout.memref.rank))
+    )
+    buffer_attr = ttnn.ir.BufferTypeAttr.maybe_downcast(layout.memref.memory_space)
+    result.append(
+        graph_builder.KeyValue(
+            key="memref_memory_space", value=str(ttnn.BufferType(buffer_attr.value))
+        )
+    )
+    return result
+
+
+class OpHandler:
+    def __init__(self, op):
+        self.op = op
+        self.attrs = []
+
+    def get_id(self, names: defaultdict):
+        name = get_loc_str(self.op.location)
+        name_num = names[name]
+        id = name + "__" + str(name_num)
+        names[name] += 1
+        return id
+
+    @staticmethod
+    def get_namespace(op):
+        name = get_loc_str(op.location)
+        if op.parent and op.parent.name != "builtin.module":
+            return OpHandler.get_namespace(op.parent) + "/" + name
+        return name
+
+    def get_attributes(self):
+        # Parse Op Attributes themselves
+        for attr in self.op.attributes:
+            self.attrs.extend(AttrHandler.parse_attr(attr))
+
+    def make_graph_node(self, name_dict):
+        self.get_attributes()
+        return graph_builder.GraphNode(
+            id=self.get_id(name_dict),
+            label=self.op.name,
+            namespace=OpHandler.get_namespace(self.op),
+            attrs=self.attrs,
+        )
+
+    def make_constant_node(self, name_dict, constant_name):
+        return graph_builder.GraphNode(
+            id=self.get_id(name_dict),
+            label=constant_name,
+            namespace=OpHandler.get_namespace(self.op),
+        )
+
+
+EMPTY_OPS = [
+    "ttnn.empty",
+    "tensor.empty",
+]
+
+FILTERED_OPS = [
+    "ttnn.deallocate",
+    "ttnn.get_device",
+]
+
+
+def build_graph(module):
+    name_dict = defaultdict(int)
+    output_connections = defaultdict(int)
+    graph = graph_builder.Graph(id="tt-graph")
+
+    op_to_graph_node = {}
+
+    module_op = OpHandler(module.operation)
+    graph.nodes.append(module_op.make_graph_node(name_dict))
+
+    for op in module.body.operations:
+        append_later = []
+        for region in op.regions:
+            for block in region.blocks:
+                for op in block.operations:
+                    # Create all the nodes and constants in the first pass.
+                    operation = OpHandler(op)
+                    graph_node = operation.make_graph_node(name_dict)
+
+                    if op.name in EMPTY_OPS:
+                        append_later.append(graph_node)
+                    elif op.name not in FILTERED_OPS:
+                        graph.nodes.append(graph_node)
+
+                    op_to_graph_node[op] = graph_node
+
+                    for operand in op.operands:
+                        if operand.owner == block and operand not in op_to_graph_node:
```

Comment:
Do we have a better way to check this now?

Reply:
I didn't really notice this as an issue so I didn't try to find a better way to check. For what it's worth, it seems to be working with TTNN modules...

Reply:
So `if operand.owner == block` checks if the operand is an arg instead of an operation. So i was wondering if we have a cleaner way to check this with all the pybind changes that happened in the meantime?

Reply:
Yes, I see now that you're talking about values as opposed to ops, we can check this with `isinstance(operand, ir.Value)`, I'll add that in the next change.